File: golang/uPIMulator/src/simulator/dpu/reg/condition_reg.go
package reg

import (
	"errors"
	"uPIMulator/src/linker/kernel/instruction/cc"
)

type ConditionReg struct {
	conditions map[cc.Condition]bool
}

func (this *ConditionReg) Init() {
	this.conditions = make(map[cc.Condition]bool, 0)

	this.ClearConditions()
}

func (this *ConditionReg) Fini() {
}

func (this *ConditionReg) Condition(condition cc.Condition) bool {
	if condition == cc.TRUE {
		return true
	} else if condition == cc.FALSE {
		return false
	} else {
		return this.conditions[condition]
	}
}

func (this *ConditionReg) SetCondition(condition cc.Condition) {
	if condition == cc.TRUE || condition == cc.FALSE {
		err := errors.New("condition is true or false")
		panic(err)
	}

	this.conditions[condition] = true
}

func (this *ConditionReg) ClearCondition(condition cc.Condition) {
	if condition == cc.TRUE || condition == cc.FALSE {
		err := errors.New("condition is true or false")
		panic(err)
	}

	this.conditions[condition] = false
}

func (this *ConditionReg) ClearConditions() {
	for i := 0; i <= int(cc.LARGE); i++ {
		condition := cc.Condition(i)

		if condition == cc.TRUE || condition == cc.FALSE {
			continue
		} else {
			this.conditions[condition] = false
		}
	}
}

File: python_cpp/uPIMulator_backend/src/simulator/reg/condition_reg.cc
#include "simulator/reg/condition_reg.h"

#include <cassert>

namespace upmem_sim::simulator::reg {

bool ConditionReg::condition(abi::isa::Condition condition) {
  if (condition == abi::isa::TRUE) {
    return true;
  } else if (condition == abi::isa::FALSE) {
    return false;
  } else {
    return bits_[condition];
  }
}

void ConditionReg::set_condition(abi::isa::Condition condition) {
  assert(condition != abi::isa::TRUE and condition != abi::isa::FALSE);
  bits_[condition] = true;
}

void ConditionReg::clear_condition(abi::isa::Condition condition) {
  assert(condition != abi::isa::TRUE and condition != abi::isa::FALSE);
  bits_[condition] = false;
}

void ConditionReg::clear_conditions() {
  for (abi::isa::Condition condition = abi::isa::TRUE;
       condition <= abi::isa::LARGE;
       condition = static_cast<abi::isa::Condition>(condition + 1)) {
    if (condition == abi::isa::TRUE or condition == abi::isa::FALSE) {
      continue;
    } else {
      clear_condition(condition);
    }
  }
}

}  // namespace upmem_sim::simulator::reg
